crates/maybenot/src/time.rs

```diff
@@ -0,0 +1,61 @@
+use std::ops::AddAssign;
+
+/// Trait representing instants in time. Allows using maybenot frameworks with
+/// custom time sources. If you want to use maybenot with a different time source
+/// than `std::time::Instant`, implement this trait for your instant type, and the
+/// [`Duration`] trait for your corresponding duration type.
+pub trait Instant: Clone + Copy {
+    type Duration: Duration;
+
+    /// Returns the amount of time elapsed from another instant to this one.
+    ///
+    /// Should return a zero duration if `earlier` is later than `self`
+    fn duration_since(&self, earlier: Self) -> Self::Duration;
+}
+
+pub trait Duration: Clone + Copy + AddAssign + PartialOrd {
+    /// Creates a new duration, spanning no time.
+    fn zero() -> Self;
+
+    /// Creates a new duration from the specified number of microseconds.
+    fn from_micros(micros: u64) -> Self;
+
+    /// Returns the total number of whole microseconds contained by this duration.
+    fn as_micros(&self) -> u128;
```

Comment:
I don't know the full codebase, but I hope we're not actually computing anything with 128-bit values: that can be quite slow, especially on 32-bit platforms.

Using a `u64` for microseconds is sufficient to represent 584558 years.  Does maybenot actually need to represent intervals longer than that?

Reply:
I just copied the API of `std::time::Duration`. I agree it's unfortunate than `from_micros` and `as_micros` use different types. But that's because they were added at different times. Were both added today both would use `u128` I'm sure. But we can of course limit it to `u64` in `maybenot` if we know we won't need larger durations. Then we need to decide what the implementation on `std::time::Duration` should behave, just cast and truncate?

Reply:
Maybenot will be fine with `u64` for sure (we max clamp all values for durations s-is).